ros/src/computing/perception/localization/packages/orb_localizer/nodes/place_recognizer/place_recognizer.cpp

```diff
@@ -0,0 +1,395 @@
+/*
+ * place_recognizer.cpp
+ *
+ *  Created on: May 18, 2017
+ *      Author: sujiwo
+ */
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <ros/ros.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <sensor_msgs/Image.h>
+#include <sensor_msgs/image_encodings.h>
+#include <cv_bridge/cv_bridge.h>
+#include <image_transport/image_transport.h>
+#include <boost/filesystem.hpp>
+#include <tf/tf.h>
+#include "System.h"
+#include "Map.h"
+#include "Frame.h"
+#include "ORBmatcher.h"
+#include "MapPoint.h"
+#include "PnPsolver.h"
+#include "ORBVocabulary.h"
+#include "ORBextractor.h"
+#include "Optimizer.h"
+#include "../common.h"
+
+
+using namespace std;
+using namespace ORB_SLAM2;
+namespace enc = sensor_msgs::image_encodings;
+
+
+
+//struct RecognizerOutput {
+//	KeyFrame *kf;
+//	geometry_msgs::PoseWithCovariance keyframePose;
+//	double imageTimestamp;
+//};
+
+
+const string orbGenericVocabFile = ORB_SLAM_VOCABULARY;
+
+// ROS Publishers
+image_transport::Publisher recognizerImageDebug;
+ros::Publisher vehiclePosePub;
+
+cv::FileStorage sysConfig;
+KeyFrameDatabase *keyframeDB;
+ORBVocabulary *keyVocab;
+ORBextractor *orbExtractor;
+ORB_SLAM2::Map *sourceMap;
+
+// Camera Parameters
+cv::Mat CameraParam;
+
+// Distortion Coefficients
+cv::Mat DistCoef;
+
+// ROS Launch Parameters
+string
+	mapPath
+		("/data/Meidai2017/logging/2016-12-26-13-21-10/map/orbx.map"),
+	configFile
+		("/data/Meidai2017/ros/etc/orb-slam2.yaml"),
+	imageTopic
+		("/camera1/image_raw");
+
+
+string kfImageDir;
+cv::Mat PostRecognition (Frame &frame, KeyFrame *kf)
+{
+	string imagePath = kfImageDir + "/" + std::to_string(kf->mnId) + ".jpg";
+	cv::Mat kfImage = cv::imread (imagePath, CV_LOAD_IMAGE_GRAYSCALE);
+	if (kfImage.empty())
+		throw std::invalid_argument("KeyFrame image not found: " + std::to_string(kf->mnId));
+
+	return kfImage;
+}
+
+
+// We eliminate full initialization of SLAM System
+// XXX: No exception handling here
+void SlamSystemPrepare (
+	const string &mapFilename,
+	const string &configPath,
+	cv::FileStorage &sconf,
+	Map **sMap,
+	ORBVocabulary **vocab,
+	KeyFrameDatabase **kfdb,
+	ORBextractor **orbext
+	)
+{
+	// open Configuration file
+	sconf = cv::FileStorage (configPath.c_str(), cv::FileStorage::READ);
+
+	// Vocabulary
+	*vocab = new ORBVocabulary ();
+	cout << endl << "Loading Custom ORB Vocabulary... " ;
+	string mapVoc = mapFilename + ".voc";
+	bool vocload = (*vocab)->loadFromTextFile (mapVoc);
+	cout << "Vocabulary loaded!" << endl << endl;
+
+	// Create KeyFrame Database
+	*kfdb = new KeyFrameDatabase(**vocab);
+
+	// Map
+	*sMap = new Map ();
+	(*sMap)->loadFromDisk(mapFilename, *kfdb, true);
```

Comment:
this is a bit risky since you throw lots of exceptions in loadFromDisk but do not handle them